risc0/zkvm/src/receipt.rs

```diff
@@ -343,56 +312,33 @@ impl SessionFlatReceipt {
     }
 }
 
-impl SegmentReceipt {
-    /// Get the [ReceiptMetadata] associated with the current receipt.
-    pub fn get_metadata(&self) -> Result<ReceiptMetadata, VerificationError> {
-        let elems = bytemuck::cast_slice(&self.seal);
-        ReceiptMetadata::decode_from_io(layout::OutBuffer(elems))
-    }
-
-    /// Verifies the integrity of this receipt.
-    ///
-    /// Uses the ZKP system to cryptographically verify that the seal does
-    /// validly indicate that this Segment was executed faithfully.
-    #[cfg(not(target_os = "zkvm"))]
-    #[must_use]
-    pub fn verify(&self) -> Result<(), VerificationError> {
-        use risc0_zkp::core::hash::sha::Sha256HashSuite;
-        let hal =
-            risc0_zkp::verify::CpuVerifyHal::<_, Sha256HashSuite<_, crate::sha::Impl>, _>::new(
-                &crate::CIRCUIT,
-            );
-        self.verify_with_hal(&hal)
+#[typetag::serde]
+impl Receipt for SegmentReceipt {
+    fn verify_with_context(&self, ctx: &VerifierContext) -> Result<(), VerificationError> {
+        use hex::FromHex;
+        let check_code = |_, control_id: &Digest| -> Result<(), VerificationError> {
+            POSEIDON_CONTROL_ID
+                .into_iter()
+                .chain(SHA256_CONTROL_ID)
+                .chain(BLAKE2B_CONTROL_ID)
+                .find(|x| Digest::from_hex(x).unwrap() == *control_id)
+                .map(|_| ())
+                .ok_or(VerificationError::ControlVerificationError)
+        };
```

Comment:
This `check_code` closure will create an iterator with all control ID's from `poseidon`, `sha256`, and `blake2b`. I think it would be more accurate if we could select the correct set of control_id's from the hash suite... I guess another question is this: is the possibility of getting the incorrect control id somehow bounded away so that it's not possible to get an accidental mismatch?

Reply:
In the near future, I'm thinking we could add this to the `VerifierContext`:
```
control_ids: HashMap<(String, String), ControlIds>
```

Where the key is the `(circuit_id, hashfn)`.

But I think this current approach is safe in the meantime, at least from a cryptographic perspective.

Reply:
Also - I think the logic in this PR means that we can simplify `check_code` to only take `&Digest` as a parameter. I think this can be done in a separate PR though